File: 백준/Silver/2667. 단지번호붙이기/단지번호붙이기.py
# 단지번호 붙이기

import sys
from collections import deque
input = sys.stdin.readline

dx = [-1, 1, 0, 0]
dy = [0, 0, -1, 1]

n = int(input())
visited = [[0]*n for _ in range(n)] # 방문 노드 기록용
array = []

count = []

def dfs(x, y):
    deq = deque()
    deq.append((x,y))
    cnt = 1
    
    while deq:
        x, y = deq.popleft() 
        visited[x][y] = 1
        
        for i in range(4):
            nx = x + dx[i]
            ny = y + dy[i]
            
            if 0 <= nx < n and 0 <= ny < n and array[nx][ny] == 1 and visited[nx][ny] == 0:
                deq.append((nx, ny))
                visited[nx][ny] = 1   
                cnt += 1

    return cnt

for i in range(n):
    array.append(list(map(int, input().rstrip())))

for i in range(n):
    for j in range(n):
        if visited[i][j] == 0 and array[i][j] == 1:
            count.append(dfs(i,j))
            
count.sort()
print(len(count))
for i in range(len(count)):
    print(count[i])

File: 백준/Silver/2667. 단지번호붙이기/단지번호붙이기.java
import java.util.*;
import java.io.*;

public class Main {

    static int n;
    static int sum;

    static int[][] graph;
    static boolean[][] visited;
    static ArrayList<Integer> answer; 

    static int[] dx = {1, 0, -1, 0}; // 우상좌하
    static int[] dy = {0, -1 ,0, 1};

    static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    static BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
    public static void main(String[] args) throws IOException {
        
        StringTokenizer st = new StringTokenizer(br.readLine());
        
        n = Integer.parseInt(st.nextToken());
        
        graph = new int[n][n];
        visited = new boolean[n][n];
        answer = new ArrayList<>();
        for(int i=0; i<n; i++){
            st = new StringTokenizer(br.readLine());
            String str = st.nextToken();

            for(int j=0; j<n; j++){
                graph[i][j] = Integer.parseInt(str.substring(j, j+1));
            }
        }

        for(int i=0; i<n; i++){
            for(int j=0; j<n; j++){
                if(visited[i][j] == false && graph[i][j] == 1){
                    visited[i][j] = true;
                    sum = 1;
                    dfs(i, j);
                    answer.add(sum);
                }
            }
        }

        Collections.sort(answer);

        bw.write(answer.size() + "\n");
        for(int i=0; i<answer.size(); i++){
            bw.write(answer.get(i) + "\n");
        }
        bw.flush();
        bw.close();
    }

    public static void dfs(int x, int y) throws IOException {

        for(int i=0; i<4; i++){
            int nx = x + dx[i];
            int ny = y + dy[i];

            // 범위 안이고, 방문하지 않으면
            if(0<= nx && nx < n && 0 <= ny && ny <n && visited[nx][ny] == false){ 
                if(graph[nx][ny] == 1){
                    sum += 1;
                    visited[nx][ny] = true;
                    dfs(nx, ny);
                }
            }
        }
    }
}
